### DISCUSSION TITLE
Remote thermostat with HMC310

### DISCUSSION THREAD

**Discussion by masterx1981** [emsesp/EMS-ESP32#1835] [General]
Hi! I have an heat pump with an HMC310 Controller, seen as an RC310.
In winter i start heating with an on-off contact on i1, and works well. 
But now i want to start cooling, and seem that setting the heat pump in cooling mode, and closing the i1 contact doesn't work. I've tried to connect an ems thermostat, and in this way the heat pump start to cool the water.
I think that the heat pump need a remote temp sensor to start to cool.
So, i've tried to configure the virtual thermostat sensor.
I've created 2 schedules:
![Screenshot_20240702_223231_Chrome](https://github.com/emsesp/EMS-ESP32/assets/76602320/d9e90fe8-25b1-4b6b-93ca-f7735f6865d6)
Then i've set the device from rc310 to rc100h.
Now in the devices i see the the RC100H device, with an entity:
Hc1 room temperature from remote.
The entity is readonly.
I want to update the temperature with the current temp in the house via ha (i have several temp sensors connected to the ha).
I've tried to read the documentation but i'm a bit confused on how to do it.
Any help?
Thanks!

**Comment by MichaelDvP** [comment]
See the wiki and #1287 
If you use HA with some external sensors, you dont need a schedule, just send the current roomtemperture via HA to `thermostat/hc1/remotetemp`, You don't need frequent updates (if you set a timeout in application-settings you need updates in this time).
Before you start sending the roomtemp you have to tell the master thermostat that there is a remote (on the device or via `thermostat/control`):

**Comment by masterx1981** [comment]
Sorry but i had difficulties understanding how it works from the posts and the "Special functions" page.
So i not need both the schedules?
I need only to set the "hc1 control device" (as named on the web page) to rc100h, and set the remotetemp.
What is the actual name on the web page?
hc1 temporary set temperature from remote 
Or
hc1 room temperature from remote 
Thanks!

**Reply by MichaelDvP** [reply]
You have to set thermostat/remotetemp from your HA system, there is no entity for it, it's just a command for automatisation and the value only show up after it is set. (You can use the scheduler or telnet to trigger HA entity creation  first time if you need this.)

@proddy should we change remotetemp to `CMD` to show in web page? This will always create a HA entity. (CMD is now only enum type in HA, but we can use UOM:degees to set it to number.

**Reply by proddy** [reply]
@MichaelDvP yes, let's create remotetemp as a CMD. The CMD is a nice touch and we should use it more anyway. If you do the back-end I can take a look at the Web and also change the HA code to make it work.

**Reply by masterx1981** [reply]
I've seen present in the changelog of firmware dev 26 the
"thermostat/remotetemp as command https://github.com/emsesp/EMS-ESP32/discussions/1835"
I'm supposed to see something in the web page, or on the mqtt entities? By now i'm not able to see how to call the command....
Thanks!

**Reply by masterx1981** [reply]
Maybe i've understood: now the "hc1 room temperature from remote" and "hc1 room humidity from remote" call the command that set the temperature, and automatically enable the remote thermostat.
After the last reboot (after installing the dev 26) the system have automatically enabled the remote thermostat without doing anything more than setting the remote temperature (scheduled from HA every 5 minutes).
Thanks!

**Comment by masterx1981** [comment]
So, for now i can issue a thermostat/remotetemp at startup via scheduler, then update the entity from HA?
By now i'm updating the "hc1 room temperature from remote" entity on HA, and the lcd of the heat pump show correctly the temperature set of the room, so i think that is working (and also the heat pump start cooling, as i hoped).

**Comment by thombec** [comment]
Hi,

I am currently working at the same topic - how to manual start the coolling function at the Buderus/Bosch heatpump. 

I am using a WPLS.2. Here it is neccessary to use a RC100H controller to check the humidity an to calculate the dew point. In addition humidity sensors at the surface of the pipes at this type are neccessary.

I recognized, that the cooling mode will start, when the damped outsidetemperature reached the programmed value from the HMC310 called "Kühlbetrieb ab". Due to the fact that there is a programmed delay in turnig on this function (>=1h) the cooling functionallyity will start min. 1h after the damped outside temperature reached. Currently this can only programmed directly at HMC310 - the delays entities like for heating (hc1_heat_off_delay...) I cant find in EMS-ESP. When this starts the roomtemperature will be checked again the programmed tempearture. If this is above the vents of the underfloor heating will open 30 min later. If now the temperature of the buffer storage / "Vorlauftemperatur" is above the programmed room temperature the heat pump start to cool. The minimum setpoint here is the minimum flow temperature (hc1_hp_min_flow_temp). The programmed value dew point offset (hc1_dew_point_offset) will check, that the calculated dew point from RC100H wont reached while cooling down to the programmed min temperature hc1_hp_min_flow_temp.

Now my target is to find a way to control the heatpump so I can modify the startpoint of the cooling procedure. My first experience here was, that some of the needed entities are currently not visible in EMS-ESP. So the programming at home assistant wont work (so easy).  I hope, that we can find a way to figure out these values and make them reachable in homeassistant.

**Reply by MichaelDvP** [reply]
for the mssing values follow https://emsesp.github.io/docs/FAQ/#decoding-ems-telegrams
log all, change value on the HMC310, check log for single value change published by HMC310, try this tellegram/offset as custom entity.

**Reply by thombec** [reply]
thank you for information.

I replied here:

https://github.com/emsesp/EMS-ESP32/issues/1857

**Comment by masterx1981** [comment]
On this subject, by now i'm sending the thermostat/remotetemp command via a timer schedule, at 00:00 (once at startup), then the HA instance update the value every 5 minutes with the house temperature. But seem that sometime the ems esp send the command too soon, and the heat pump doesn't get it.
I've not understood if there is a way so that i can send the command via mqtt/ha....
Thanks!

**Reply by proddy** [reply]
yes you can send commands via HA or MQTT.

I'm thinking, @MichaelDvP, shall we extend the Scheduler so you can execute the event? Even if its disabled. That way you could create quite complex commands and conditions, keep it disabled and invoke it manually via the API.

**Reply by MichaelDvP** [reply]
Something like a new type of schedule: `single action`: inputs just activate, name, command and value: executes one time after activation and sets active to false. Activate in the dialog means execution on ok7update button, in the list it is always deactive. Or we add a button for execution ad skip the checkbox.
But you can also do this now: use trigger onChange with a custom value. With changing the custom value by command you'll trigger the schedule. Or use condition `1`,but then you have to send on and off withing 10 sec to have a one time execution.

**Reply by MichaelDvP** [reply]
The problem with sending remotetemp at startup is, that the thermostat and hc is detected much later. Te startup command makes some retries, but if it takes too long to detect the hc the command is not executet. With the change to CMD there is no need to use the startup scheduler to make remotetemp show up, the command is always generating a HA entity.

**Reply by proddy** [reply]
> Something like a new type of schedule: `single action`: inputs just activate, name, command and value: executes one time after activation and sets active to false. Activate in the dialog means execution on ok7update button, in the list it is always deactive. Or we add a button for execution ad skip the checkbox. 

I'll turn this into a separate discussion, I think we need some more use cases before implementing it.

> But you can also do this now: use trigger onChange with a custom value. With changing the custom value by command you'll trigger the schedule. Or use condition `1`,but then you have to send on and off withing 10 sec to have a one time execution.

yes! I'll add this to the wiki.

**Comment by mheyse** [comment]
A question regarding CMD: the modbus integration currently does not write CMDs because it needs to know the data type size, and now that remotetemp/remotehum are CMDs setting them via Modbus does not work any more.

What data type are CMDs? Is it always 16 bits (INT16/UINT16)?

**Reply by MichaelDvP** [reply]
CMD does not have a value. You can only send to ems-esp, which is always treated as a char string. The specific command function scans char-array and converts to needed value for the bus.

**Reply by mheyse** [reply]
No value is stored, understood, but a value is still transmitted to ems-esp that needs to be encoded in a specific number of bytes for transmission via Modbus (which is not a problem for MQTT because it just uses strings). If the data type is unknown it's difficult to know how many bytes to allocate. So we could either default to always using 32 bits (which breaks backward compatibility, because remotetemp and remotehum currently are 16 bit values) or try to guess a sensible register size from the entities min, max and numeric_operator values when creating the entity/register mapping (not sure if this always works).

**Reply by mheyse** [reply]
I implemented the "guess a register size" approach, and for all currently existing CMDs this yields the same register sizes as before, so I think all is good.

**Reply by proddy** [reply]
in dev-32
